pkg/sql/schemachanger/scbuild/internal/scbuildstmt/partition_zone_config.go

```diff
@@ -0,0 +1,385 @@
+	if err := finalZone.ValidateTandemFields(); err != nil {
+		err = errors.Wrap(err, "could not validate zone config")
+		err = pgerror.WithCandidateCode(err, pgcode.InvalidParameterValue)
+		err = errors.WithHint(err,
+			"try ALTER ... CONFIGURE ZONE USING <field_name> = COPY FROM PARENT [, ...] to "+
+				"populate the field")
+		return err
+	}
+	pzo.setZoneConfigToWrite(partialZone)
+	return err
+}
+
+// panicIfNoPartitionExistsOnIdx panics if the partition referenced in a
+// ALTER PARTITION ... OF TABLE does not exist on the provided index.
+func (pzo *partitionZoneConfigObj) panicIfNoPartitionExistsOnIdx(
+	b BuildCtx, n *tree.SetZoneConfig,
+) {
+	zs := n.ZoneSpecifier
+	// If we allow StarIndex to be set in the DSC, we will have to guard against
+	// that here as well.
+	if zs.TargetsPartition() && len(zs.TableOrIndex.Index) != 0 {
+		partitionName := string(zs.Partition)
+		var indexes []scpb.IndexPartitioning
+		idxPart := b.QueryByID(pzo.tableID).FilterIndexPartitioning()
+		idxPart.ForEach(func(_ scpb.Status, _ scpb.TargetStatus, e *scpb.IndexPartitioning) {
+			if e.IndexID != pzo.indexID {
+				return
+			}
+			partition := tabledesc.NewPartitioning(&e.PartitioningDescriptor)
+			partition = partition.FindPartitionByName(partitionName)
+			if partition != nil {
+				indexes = append(indexes, *e)
+			}
+		})
+
+		indexName := string(zs.TableOrIndex.Index)
+		switch len(indexes) {
+		case 0:
+			panic(fmt.Errorf("partition %q does not exist on index %q", partitionName, indexName))
+		case 1:
+			return
+		default:
+			panic(errors.AssertionFailedf("partition %q exists multiple times on index %q",
+				partitionName, indexName))
+		}
+	}
+}
+
+// panicIfBadPartitionReference panics if the partition referenced in a
+// ALTER PARTITION ... OF TABLE does not exist or if it exists on multiple
+// indexes. Otherwise, we find the existing index and save in to our AST.
+// In cases where we find the partition existing on two indexes, we check
+// to ensure that we are not in a backfill case before panicking.
+func (pzo *partitionZoneConfigObj) panicIfBadPartitionReference(b BuildCtx, n *tree.SetZoneConfig) {
+	zs := &n.ZoneSpecifier
+	// Backward compatibility for ALTER PARTITION ... OF TABLE. Determine which
+	// index has the specified partition.
+	//
+	// If we allow StarIndex to be set in the DSC, we will have to guard against
+	// that here as well.
+	if zs.TargetsPartition() && len(zs.TableOrIndex.Index) == 0 {
+		partitionName := string(zs.Partition)
+
+		// TODO(before merge): [reviewer callout] can we guarantee that this will
+		// be in the same order as NonDropIndexes()? Mainly for case 2 in the
+		// switch case below. My understanding is yes, but I want to confirm.
+		var indexes []scpb.IndexPartitioning
+		idxPart := b.QueryByID(pzo.tableID).FilterIndexPartitioning()
+		idxPart.ForEach(func(_ scpb.Status, _ scpb.TargetStatus, e *scpb.IndexPartitioning) {
```

Comment:
Do we need to look at the current status or target here? This will include dropping objects as well right now.